In [2]:
import numpy as np
from scipy.linalg import cho_factor, cho_solve

Cholesky decompostion facrs a symmetric matrix into a lower triangular matrix L and its transpose such that

$$
A = LL^\dagger.
$$

For example,

\begin{pmatrix}
4.31  &  0.094 & 1.599 & 0.634 & -0.16 \\
0.094 &  6.961 & 1.984 &  3.656 & -3.814 \\
1.599 &  1.984 & 7.412 & -1.473 & -2.258 \\
0.634 &  3.656 & -1.473 &  4.386 & -1.518 \\
0.16  & -3.814 & -2.258 & -1.518 &  2.553
\end{pmatrix}

decomposes into the product of

\begin{pmatrix}
2.076 & 0 &    0  &   0  &   0 \\
0.045 &  2.638  0 &    0 &   0 \\
0.77  &  0.739 & 2.505 & 0 &   0 \\
0.305 &  1.381 &-1.089 & 1.095 & 0 \\
-0.077 &  -1.445 & -0.452 &  0.007 & 0.506
\end{pmatrix}

and

\begin{pmatrix}
2.076 & 0.045 & 0.77 &  0.305 & -0.077 \\
0 &    2.638 & 0.739 &  1.381 & -1.445 \\
0 &    0 &    2.505 &  -1.089 & -0.452 \\
0 &    0 &    0  &   1.095 & 0.007 \\
0 &    0 &    0  &   0  &   0.506
 \end{pmatrix}

In [12]:
np.random.seed(1234)
A = np.random.normal(size = (5, 5))
A = A @ A.T

with np.printoptions(precision=3, suppress=True):
    print(A)

print()
L = np.linalg.cholesky(A)
#print(L)

with np.printoptions(precision=3, suppress=True):
    print( L @ L.T)
    
b = np.random.normal(size = (5, 1))
print(b)

[[ 4.31   0.094  1.599  0.634 -0.16 ]
 [ 0.094  6.961  1.984  3.656 -3.814]
 [ 1.599  1.984  7.412 -1.473 -2.258]
 [ 0.634  3.656 -1.473  4.386 -1.518]
 [-0.16  -3.814 -2.258 -1.518  2.553]]

[[ 4.31   0.094  1.599  0.634 -0.16 ]
 [ 0.094  6.961  1.984  3.656 -3.814]
 [ 1.599  1.984  7.412 -1.473 -2.258]
 [ 0.634  3.656 -1.473  4.386 -1.518]
 [-0.16  -3.814 -2.258 -1.518  2.553]]
[[-0.46930528]
 [ 0.67555409]
 [-1.81702723]
 [-0.18310854]
 [ 1.05896919]]


Let's solve the linear system

$$
\begin{pmatrix}
4.31  &  0.094 & 1.599 & 0.634 & -0.16 \\
0.094 &  6.961 & 1.984 &  3.656 & -3.814 \\
1.599 &  1.984 & 7.412 & -1.473 & -2.258 \\
0.634 &  3.656 & -1.473 &  4.386 & -1.518 \\
0.16  & -3.814 & -2.258 & -1.518 &  2.553
\end{pmatrix}
\begin{pmatrix}
    x_1 \\
    x_2 \\
    x_3 \\
    x_4 \\
    x_5
\end{pmatrix}
= 
\begin{pmatrix}
    -0.4693 \\
    0.6756 \\
    -1.8170 \\
    -0.1831 \\
    1.0590
\end{pmatrix}
$$

We do our Cholesky decompostion and write

$$
\begin{pmatrix}
2.076 & 0 &    0  &   0  &   0 \\
0.045 &  2.638  0 &    0 &   0 \\
0.77  &  0.739 & 2.505 & 0 &   0 \\
0.305 &  1.381 &-1.089 & 1.095 & 0 \\
-0.077 &  -1.445 & -0.452 &  0.007 & 0.506
\end{pmatrix}
\underbrace{
    \begin{pmatrix}
    2.076 & 0.045 & 0.77 &  0.305 & -0.077 \\
    0 &    2.638 & 0.739 &  1.381 & -1.445 \\
    0 &    0 &    2.505 &  -1.089 & -0.452 \\
    0 &    0 &    0  &   1.095 & 0.007 \\
    0 &    0 &    0  &   0  &   0.506
    \end{pmatrix}
    \begin{pmatrix}
        x_1 \\
        x_2 \\
        x_3 \\
        x_4 \\
        x_5
    \end{pmatrix}
}_{\tilde{x}}
= 
\begin{pmatrix}
    -0.4693 \\
    0.6756 \\
    -1.8170 \\
    -0.1831 \\
    1.0590
\end{pmatrix}
$$

Now we can write,

$$
\begin{pmatrix}
2.076 & 0 &    0  &   0  &   0 \\
0.045 &  2.638  0 &    0 &   0 \\
0.77  &  0.739 & 2.505 & 0 &   0 \\
0.305 &  1.381 &-1.089 & 1.095 & 0 \\
-0.077 &  -1.445 & -0.452 &  0.007 & 0.506
\end{pmatrix}
\begin{pmatrix}
    \tilde{x_1} \\
    \tilde{x_2} \\
    \tilde{x_3} \\
    \tilde{x_4} \\
    \tilde{x_5}
\end{pmatrix}
=
\begin{pmatrix}
    -0.4693 \\
    0.6756 \\
    -1.8170 \\
    -0.1831 \\
    1.0590
\end{pmatrix}
$$

We can easily solve the above equation for $\tilde{x}$ by forward substitution.
Then we can get our original unkwon vector by backward substitution,

$$
\begin{pmatrix}
    2.076 & 0.045 & 0.77 &  0.305 & -0.077 \\
    0 &    2.638 & 0.739 &  1.381 & -1.445 \\
    0 &    0 &    2.505 &  -1.089 & -0.452 \\
    0 &    0 &    0  &   1.095 & 0.007 \\
    0 &    0 &    0  &   0  &   0.506
\end{pmatrix}
\begin{pmatrix}
    x_1 \\
    x_2 \\
    x_3 \\
    x_4 \\
    x_5
\end{pmatrix}
= 
\begin{pmatrix}
    \tilde{x_1} \\
    \tilde{x_2} \\
    \tilde{x_3} \\
    \tilde{x_4} \\
    \tilde{x_5}
\end{pmatrix}
$$

In [16]:
x1 = np.linalg.solve(A, b)
print(x1)
print()

L = cho_factor(A, lower = True)
x2 = cho_solve(L, b)
print(x2)

np.allclose(x1, x2)

[[ 0.14191414]
 [ 3.00959324]
 [ 0.00698134]
 [-1.0868875 ]
 [ 4.28058344]]

[[ 0.14191414]
 [ 3.00959324]
 [ 0.00698134]
 [-1.0868875 ]
 [ 4.28058344]]


True

In [21]:
np.random.seed(18)

X = np.random.normal(size = (1000000, ))
Y = np.random.normal(size = (1000000, ))

#np.cov(X, Y)

C = np.array([ [1, 0.5], [0.5, 1] ])
#print(C)

X.shape = (1, X.size)
Y.shape = (1, Y.size)

XY = np.vstack((X, Y))

L = np.linalg.cholesky(C)

UV = L @ XY

U = UV[0,:]
V = UV[1, :]

print(np.cov(U, V))

[[0.99968938 0.49948667]
 [0.49948667 0.99936442]]


The covariance of a random variable $X$ and $Y$ (mean - and variance 1) is given by, 

$$
\mbox{cov}(X, Y) = \mathbb{E}(XY^\dagger).
$$

So if we have a collection of uncorrelated, noramlly distributed random variables,

$$
\mbox{cov}(X_i, X_j) = \delta_{i,j} = I
$$

Suppose we have a set of correlated variables to covariance matrix $C$.  We can do a Cholesky decomposition and write $C = LL^\dagger$.

Suppose we have two random variables in $V$ and $U$ is defined such that $U = LV$.  What is the covariance?


$$
\mathbb{E}(UU^\dagger) = \mathbb{E}( (LV)(LV)^\dagger) ) = \mathbb{E}(LVV^\dagger L^\dagger)
$$
We can factor out $L$,

$$
L\mathbb{E}(VV^\dagger)L^\dagger = LIL^\dagger = LL^\dagger = C
$$